drivers/gpio/hcsr04_driver.go

```diff
@@ -0,0 +1,183 @@
+	echoPin    *gpio.DirectPinDriver
+	// triggerPin             rpio.Pin
+	// echoPin                rpio.Pin
+	mux                    sync.Mutex
+	Measure                float32 // The last measure
+	distanceMonitorControl chan int
+	distanceMonitorStarted bool
+	gobot.Commander
+}
+
+// NewHCSR04 creates a new HCSR04 instance
+func NewHCSR04(a gobot.Adaptor, triggerPinID string, echoPinID string) *HCSR04 {
+	hcsr04 := &HCSR04{
+		name:       gobot.DefaultName("HCSR04"),
+		triggerPin: gpio.NewDirectPinDriver(a, triggerPinID),
+		echoPin:    gpio.NewDirectPinDriver(a, echoPinID),
+		connection: a,
+		Commander:  gobot.NewCommander(),
+	}
+
+	// hcsr04 := HCSR04{
+	// 	triggerPinID: triggerPinID,
+	// 	echoPinID:    echoPinID,
+	// }
+	// hcsr04.triggerPin = rpio.Pin(hcsr04.triggerPinID)
+	// hcsr04.triggerPin.Mode(rpio.Output)
+
+	// hcsr04.echoPin = rpio.Pin(hcsr04.echoPinID)
+	// hcsr04.echoPin.Mode(rpio.Input)
+	// hcsr04.echoPin.PullDown()
+	// hcsr04.triggerPin.Low()
+	return hcsr04
+}
+
+// MeasureDistance measure the distance in front of sensor in meters
+// and returns the measure
+// MeasureDistance triggers a distance measurement by the sensor
+//
+// ! MeasureDistance is not design to work in a fast loop
+// For this specific usage, use StartDistanceMonitor associated with GetDistance Instead
+func (hcsr04 *HCSR04) MeasureDistance() (float32, error) {
+	hcsr04.mux.Lock()
+	defer hcsr04.mux.Unlock()
+	pulseDuration, err := hcsr04.measurePulse()
+	if err != nil {
+		return 0, err
+	}
+	hcsr04.Measure = pulseToDistance(pulseDuration)
+	return hcsr04.Measure, nil
+}
+
+func (hcsr04 *HCSR04) emitTrigger() {
+	hcsr04.triggerPin.On()
+	time.Sleep(10 * time.Microsecond)
+	hcsr04.triggerPin.Off()
+}
+
+func (hcsr04 *HCSR04) measurePulse() (int64, error) {
+	startChan := make(chan int64)
+	stopChan := make(chan int64)
+	startQuit := false
+	stopQuit := false
+	var startTime int64
+	var stopTime int64
+	go getPinStateChangeTime(hcsr04.echoPin, 1, startChan, &startQuit)
+	hcsr04.emitTrigger()
+	readedValue, _ := hcsr04.echoPin.DigitalRead()
+	if readedValue == 1 {
+		return 0, errors.New("already receiving echo")
+	}
+	select {
+	case t := <-startChan:
+		startTime = t
+	case <-time.After(measurementCycle * time.Millisecond):
+		startQuit = true
+		return 0, fmt.Errorf("echo not received after %d milliseconds", measurementCycle)
+	}
+	go getPinStateChangeTime(hcsr04.echoPin, 0, stopChan, &stopQuit)
+	select {
+	case t := <-stopChan:
+		stopTime = t
+	case <-time.After(measurementCycle * time.Millisecond):
+		stopQuit = true
+		return 0, fmt.Errorf("echo received for more than %d milliseconds", measurementCycle)
+	}
+	return stopTime - startTime, nil
+}
+
+func getPinStateChangeTime(pin *gpio.DirectPinDriver, state int, outChan chan int64, quit *bool) {
+
+	for {
+		// readedValue, _ := pin.DigitalRead()
+		readedValue := 1
+		// fmt.Println("Lectura: ", readedValue, state, *quit)
+		if readedValue != state && !*quit {
+			break
+		}
+	}
```

Comment:
this loop waits for "an ongoing measurement is finished", so it will possibly not work, if the pin.DigitalRead() is substituted by a constant of "1" (especially for the first call)